src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -0,0 +1,443 @@
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'total number of entries' field");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_CD_START_OFFSET, cd_offset)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'central directory size' field");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_CD_SIZE_OFFSET, cd_size)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'central directory offset' field");
+		return false;
+	}
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_ensure_valid_params (uint8_t* buf, size_t buf_len, size_t index, size_t to_read)
+{
+	assert (buf != nullptr);
+	if (index + to_read > buf_len) {
+		log_fatal (LOG_DEFAULT, "Buffer too short to read %u bytes of data", to_read);
+		return false;
+	}
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, uint16_t& u)
+{
+	if (!zip_ensure_valid_params (buf, buf_len, index, sizeof (u)))
+		return false;
+
+	u = static_cast<uint16_t> (buf [index + 1] << 8) |
+		static_cast<uint16_t> (buf [index]);
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, uint32_t& u)
+{
+	if (!zip_ensure_valid_params (buf, buf_len, index, sizeof (u)))
+		return false;
+
+	u = (static_cast<uint32_t> (buf [index + 3]) << 24) |
+		(static_cast<uint32_t> (buf [index + 2]) << 16) |
+		(static_cast<uint32_t> (buf [index + 1]) << 8)  |
+		(static_cast<uint32_t> (buf [index + 0]));
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, uint8_t (&sig)[4])
+{
+	static constexpr size_t sig_size = sizeof(sig);
+
+	if (!zip_ensure_valid_params (buf, buf_len, index, sig_size))
+		return false;
+
+	memcpy (sig, buf + index, sig_size);
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, size_t count, char*& characters)
+{
+	if (!zip_ensure_valid_params (buf, buf_len, index, count))
+		return false;
+
+	characters = new char[count + 1];
+	memcpy (characters, buf + index, count);
+	characters [count] = '\0';
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_entry_info (uint8_t* buf, size_t buf_len, size_t& buf_offset, uint16_t& compression_method, uint32_t& local_header_offset, uint32_t& file_size, char*& file_name)
+{
+	static constexpr size_t CD_COMPRESSION_METHOD_OFFSET = 10;
+	static constexpr size_t CD_UNCOMPRESSED_SIZE_OFFSET  = 24;
+	static constexpr size_t CD_FILENAME_LENGTH_OFFSET    = 28;
+	static constexpr size_t CD_EXTRA_LENGTH_OFFSET       = 30;
+	static constexpr size_t CD_LOCAL_HEADER_POS_OFFSET   = 42;
+	static constexpr size_t CD_COMMENT_LENGTH_OFFSET     = 32;
+
+	size_t index = buf_offset;
+	if (!zip_ensure_valid_params (buf, buf_len, index, ZIP_CENTRAL_LEN))
+		return false;
+
+	uint8_t signature[4];
+	if (!zip_read_field (buf, buf_len, index, signature)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Central Directory entry signature");
```

Comment:
Ditto all of these.  If *any* `log_fatal()` call is hit because something went wrong, we should just *die immediately*, not...muddle through at runtime with all of these return value checks.

Reply:
Making every `log_fatal` exit immediately without returning a boolean would require passing extra parameters to some of those functions (apk name, central directory index for instance - they aren't relevant for e.g. `zip_read_field`) which have nothing to do with what the functions do. I will modify the code to use `log_error` instead of `log_fatal` in the "inner", return a boolean and do a "final" `log_fatal` + `exit` instead.